text/0029-zero-alloc-messaging-flow.md

```diff
@@ -0,0 +1,390 @@
+- The fourth reason is to improve the interface by which messages are sent - instead of needing a Begin...() to add header information and an End...() to perform sending logic, MessageQueueContainer now simply has `public void SendMessage<TMessageType>(in TMessageType message) where TMessageType: IMessage{}`, which can handle both responsibilities, as well as message batching responsibilities, all in one location.
+
+## Message Header
+
+Each message sent will be prepended with a standardized message header, which is also represented as a struct with the following data:
+
+```c#
+internal struct MessageHeader
+{
+    // Sized 1 byte
+    public MessageType MessageType;
+    // Sized 1 byte
+    public NetworkChannel NetworkChannel
+    // Sized 2 bytes
+    public short MessageSize;
+}
+```
+
+Since the largest value is 2 bytes in size, the struct should be aligned to a 2 byte alignment, which should make the struct packed and the total header size equal to exactly 4 bytes, which can be sent and received using `WriteValue` and `ReadValue` without any wasted bandwidth.
+
+In addition to the message header, there's also a batch header:
+
+```C#
+internal struct BatchHeader
+{
+    // Sized 2 bytes
+    public ushort BatchSize;
+}
+```
+
+### Message timing
+
+Based on issues we've seen with UpdateStage recently, rather than the current support for sending messages to be executed at any time in the frame, we will change network processing to always process messages during `EarlyUpdate` and remove `UpdateStage` from both the internal and RPC interfaces.
+
+For situations where a message needs to be processed at a certain stage, the following utility class is added to be utilized on the client side rather than dictated on the server side:
+
+```C#
+public class DelayUtility
+{
+    public enum DelayStage: byte
+    {
+        PreFixedUpdate = NetworkUpdateStage.FixedUpdate,
+        PreUpdate = NetworkUpdateStage.Update,
+        PreLateUpdate = NetworkUpdateStage.PreLateUpdate,
+        PostLateUpdate = NetworkUpdateStage.PostLateUpdate
+    }
+
+    public delegate DelayedNetworkAction();
+
+    public void DelayUntil(DelayStage stage, DelayedNetworkAction callback);
+    
+    // Returns true if processing should be delayed, false if the specified stage is the current stage
+    // and thus the message should be processed.
+    public bool DelayMessage(DelayStage stage, ref FastBufferReader reader, NetworkContext context);
+}
+```
+
+During each of those three stages, we'll process any delayed queues.
+
+This uses a static queue of (configurable) fixed size per DelayStage value, so no allocations will be incurred in the queueing itself, **however, generally speaking, in the first version, creating delegates will cause an allocation.** There's no real way around that.
```

Comment:
Would it be horrid to just make DelayedNetworkAction an actual method that is always invoked and have that invoke a virtual method like `protected virtual void OnDelayedNetworkAction();` ?